src/parser.rs

```diff
@@ -0,0 +1,1120 @@
+            Token::Lt => Precedence::LessGreater,
+            Token::Gt => Precedence::LessGreater,
+            Token::Plus => Precedence::Sum,
+            Token::Minus => Precedence::Sum,
+            Token::Slash => Precedence::Product,
+            Token::Asterisk => Precedence::Product,
+            Token::Lparen => Precedence::Call,
+            Token::Lbracket => Precedence::Index,
+            _ => Precedence::Lowest,
+        }
+    }
+}
+
+pub struct Parser<'a> {
+    l: Lexer<'a>,
+
+    cur_token: Token,
+    peek_token: Token,
+}
+
+pub fn parse(input: &str) -> Result<Node, ParseErrors> {
+    let l = Lexer::new(input);
+    let mut p = Parser::new(l);
+    let prog = p.parse_program()?;
+
+    Ok(Node::Program(Box::new(prog)))
+}
+
+impl<'a> Parser<'a> {
+    pub fn new(l: Lexer) -> Parser {
+        let mut l = l;
+        let cur = l.next_token();
+        let next = l.next_token();
+        Parser {
+            l,
+            cur_token: cur,
+            peek_token: next,
+        }
+    }
+
+    pub fn parse_program(&mut self) -> Result<Program, ParseErrors> {
+        let mut prog = Program::new();
+        let mut errors = ParseErrors::new();
+        let mut tok = self.cur_token.clone();
+
+        while tok != Token::EOF {
+            match self.parse_statement() {
+                Ok(stmt) => prog.statements.push(stmt),
+                Err(err) => errors.push(err),
+            }
+            self.next_token();
+            tok = self.cur_token.clone();
+        }
+
+        if errors.len() > 0 {
+            return Err(errors);
+        }
+
+        Ok(prog)
+    }
+
+    fn parse_statement(&mut self) -> ParseResult<Statement> {
+        match &self.cur_token.clone() {
+            Token::Let => self.parse_let_statement(),
+            Token::Return => self.parse_return_statement(),
+            _ => self.parse_expression_statement(),
+        }
+    }
+
+    fn parse_expression_statement(&mut self) -> ParseResult<Statement> {
+        let expression = self.parse_expression(Precedence::Lowest)?;
+
+        if self.peek_token_is(&Token::Semicolon) {
+            self.next_token();
+        }
+
+        Ok(Statement::Expression(Box::new(ExpressionStatement { expression })))
+    }
+
+    fn parse_expression(&mut self, precedence: Precedence) -> ParseResult<Expression> {
+        let mut left_exp;
+
+        if let Some(f) = self.prefix_fn() {
+            left_exp = f(self)?;
+        } else {
+            return Err(format!("no prefix parse function for {} found", self.cur_token));
+        }
+
+        while !self.peek_token_is(&Token::Semicolon) && precedence < self.peek_precedence() {
+            match self.infix_fn() {
+                Some(f) => {
+                    self.next_token();
+                    left_exp = f(self, left_exp)?;
+                }
+                None => return Ok(left_exp),
+            }
+        }
+
+        Ok(left_exp)
+    }
+
+    fn prefix_fn(&mut self) -> Option<PrefixFn> {
```

Comment:
The way I did this felt kind of odd. He had in the book a global map that had a string key to match the token to the associated function used to parse. I got the same effect with this match and passing functions, but it felt a little weird. Is there a better way to do this?

Reply:
This is basically the more type-safe version of that. Instead of looking up the function by the token's string repr, you're looking it up by the token variant. I don't think this is weird but I'd probably manually inline `infix_fn` at its call location to make it more apparent / readable what its purpose is (so that the person reading this code doesn't have to jump around), also because the `infix_fn` is short and only used at this one location. 
So I'd write `let infix_fn = match { ...` and then `match infix_fn { ...`.